kube-runtime/src/coordination.rs

```diff
@@ -0,0 +1,603 @@
+        let event = match res {
+            Ok(event) => event,
+            Err(err) => {
+                tracing::error!(error = ?err, "error from lease watcher stream");
+                return;
+            }
+        };
+        match event {
+            Event::Applied(lease) => self.update_state(Some(lease)),
+            Event::Restarted(leases) => {
+                for lease in leases {
+                    self.update_state(Some(lease));
+                }
+            }
+            Event::Deleted(_) => {
+                tracing::warn!(
+                    "lease {}/{} unexpectedly deleted, will re-create",
+                    self.config.namespace,
+                    self.config.name
+                );
+                self.update_state(None);
+            }
+        };
+    }
+
+    /// Fetch the target lease from the API, and update observation info as needed.
+    #[tracing::instrument(level = "debug", skip_all)]
+    async fn k8s_fetch_lease_and_update(&mut self) -> Result<()> {
+        // Attempt to fetch the target lease, updating our last observed info on the lease.
+        let lease_opt = timeout(self.config.api_timeout, self.api.get_opt(&self.config.name))
+            .await
+            .map_err(|_err| Error::TimeoutError)?
+            .map_err(Error::ClientError)?;
+
+        // Lease info returned from the API, update observation info.
+        if let Some(lease) = lease_opt {
+            self.update_state(Some(lease));
+            return Ok(());
+        }
+        Ok(())
+    }
+
+    /// Attempt to acquire or renew the target lease.
+    #[tracing::instrument(level = "debug", skip_all, err)]
+    #[allow(clippy::cast_possible_truncation)]
+    async fn try_acquire_or_renew(&mut self) -> Result<()> {
+        tracing::debug!("try_acquire_or_renew");
+        // 1. Fetch the current state of the lease if following or standby.
+        if matches!(&self.state, State::Following { .. } | State::Standby) {
+            self.k8s_fetch_lease_and_update().await?;
+        }
+
+        // 2. Determine what type of update needs to be made to the lease.
+        // If following, and the lease is not expired (according to our own local time records), then no-op.
+        if matches!(&self.state, State::Following { .. }) && !self.is_lease_expired() {
+            return Ok(());
+        }
+        // Else, we need to patch the lease. Build up changeset.
+        let now = chrono::Utc::now();
+        let mut lease = self.state.get_lease().cloned().unwrap_or_default();
+        lease
+            .meta_mut()
+            .name
+            .get_or_insert_with(|| self.config.name.clone());
+        lease
+            .meta_mut()
+            .namespace
+            .get_or_insert_with(|| self.config.namespace.clone());
+        let spec = lease.spec.get_or_insert_with(Default::default);
+        spec.lease_duration_seconds = Some(self.config.lease_duration.as_secs_f32() as i32);
```

Comment:
I feel like this whole block here should be replaced with something like `unwrap_or_else(|| Lease::from(self.config)` and implement `From`.

currently it looked to me like we were trying to update the name/ns mid-operation, which should never happen. it would separate the static properties from the dynamic ones (such as `acquire_time` and `renew_time`)

Reply:
Good call. Will do.

Reply:
Ok, I think I've got a reasonable changeset for this in place.